pytroll_monitor/checkmk_logger.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2021
+
+# Author(s):
+
+#   Gerrit Holl <gerrit.holl@dwd.de>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Logger to report status to checkmk local checks.
+
+This module contains a logging handler that will gather information based on
+logging messages and report the status to a checkmk local check file.  It is
+still a work in progress.  So far it just writes a single line to the indicated
+status_line, and the only thing that changes is the initial digit, where 0 =
+OK, 1 = WARN, 2 = CRIT, and 3 = UNKNOWN.  Status starts at unknown.  When a
+message is logged with level warn, status is set to warn.  When a message is
+logged with level error or critical, status is set to critical.  When the
+message "All n files produced nominally" for n>0 is logged, status is set to
+OK.
+
+For more information on checkmk local checks, see
+https://docs.checkmk.com/latest/en/localchecks.html
+
+An example logging config illustrating how to use this::
+
+    version: 1
+    disable_existing_loggers: false
+    formatters:
+      pytroll:
+        format: '[%(asctime)s %(levelname)-8s %(name)s] %(message)s'
+    handlers:
+      console:
+        class: logging.StreamHandler
+        level: DEBUG
+        formatter: pytroll
+        stream: ext://sys.stdout
+      monitor:
+        (): pytroll_monitor.checkmk_logger.CheckMKHandler
+        level: DEBUG
+        formatter: pytroll
+        status_file: /opt/pytroll/pytroll_inst/pytroll_status
+    root:
+      level: DEBUG
+      handlers:
+        - console
+        - monitor
+
+Pass this to the logging config option for supported pytroll packages,
+such as::
+
+    satpy_launcher.py -n localhost trollflow2.yaml -c logging.yaml
+"""
+
+import logging
+
+from enum import IntEnum
+
+
+class ServiceStatus(IntEnum):
+    """Service status according to checkmk local checks.
+
+    Enumerate the service status according to checkmk local checks.
+    """
+
+    OK = 0
+    WARN = 1
+    CRIT = 2
+    UNKNOWN = 3
+
+
+class CheckMKHandler(logging.Handler):
```

Comment:
As this handler is specifically tailored for trollflow2 (with the hardcoded checks for specific log messages), could this be renamed to eg `CheckMKHandler4Trollflow2`?

Reply:
Or `Trollflow2CheckMKHandler`.

Reply:
I have renamed it now, but in principle, I was hoping to use it with the whole chain from (in our case) trollstalker, geographic gatherer, pytroll-aapp-runner, and trollflow2.  The "success" message hardcoded in this class, meaning trollflow2 has generated images, is the final step not only for trollflow2, but also for the entire chain.  However, at present only trollflow2 seems to be able to actually take a yaml logging config file.  If and when it works with other pytroll packages we can think about how to proceed.

Reply:
I think each of the steps should be monitored separately so that the failures would be spotted in the correct places.

Reply:
@pnuu Certainly, there is a lot of scope for improvement.  I see this PR as a first step giving some very rudimentary information.  I don't even know for sure yet if and how we will use it, still coordinating with our monitoring people at DWD.